ref(ui): Dynamically render contexts based on availability

jan-auer opened the pull request:
Renders contexts dynamically based on what's present in the event. This allows us to display contexts like runtimes or devices even in JavaScript events.  The logic renders 3-4 context items in the following order, but preferring those that are defined in the event (with a fallback to empty ones).

1. User
2. Browser
3. Runtime
4. OS
5. Device

So, if an event only includes Runtime and OS contexts, then this will render [User, Runtime, OS] in this exact order, where the User context is empty (read: "Unknown User").

An example for this is Electron: The context information is the same, regardless of the platform. However, events could be generated as `javascript` or `native`, which led to inconsistent context items. For node crashes, they were missing completely, and for renderer crashes we always displayed "Chrome" as browser, which is not ideal. 

<img width="1131" alt="screen shot 2018-04-16 at 16 57 35" src="https://user-images.githubusercontent.com/1433023/38819912-f4c33f0e-419c-11e8-8758-d5bd78b0fafa.png">

Previously, contexts were only shown for the following platforms, with a hard-coded set of context items:
 - **native** / **cocoa**: User, Device, OS
 - **java** (only Android!): User, Device, OS
 - **javascript**: User, Browser, OS (with a fallback to Device)
 - Then there was a dead code path for other platforms: User, Runtime, OS (with a fallback to Device)

jan-auer commented:
Maybe it still makes sense to override the preferred order based on the platform (if we decide to stick with it). For example, see [this visual diff](https://percy.io/getsentry/sentry/builds/658136/view/34309994/1280?mode=diff&snapshot=34309994).

Also, we could just hide the bar altogether, if only a single context is given, as it makes no sense to show multiple "Unknown ..." items.

billyvg commented on src/sentry/static/sentry/app/components/events/contextSummary.jsx:
```diff
@@ -133,59 +159,32 @@ class EventContextSummary extends React.Component {
+    let count = 0;
+
+    // Add defined contexts in the declared order, until we reach the limit
+    // defined by CONTEXT_COUNT_MAX.
```
Update constant name

billyvg commented on src/sentry/static/sentry/app/components/events/contextSummary.jsx:
```diff
@@ -133,59 +159,32 @@ class EventContextSummary extends React.Component {
+
+    // Add defined contexts in the declared order, until we reach the limit
+    // defined by CONTEXT_COUNT_MAX.
+    let children = KNOWN_CONTEXTS.map(({key, Component, ...props}) => {
```
I know this is from previous code but can we change the name of `children` - I think we should try to reserve it to refer to `children` from props.

billyvg commented on src/sentry/static/sentry/app/components/events/contextSummary.jsx:
```diff
@@ -133,59 +159,32 @@ class EventContextSummary extends React.Component {
-          )
-        );
-        break;
+    let count = 0;
```
nit might help if this was a more descriptive name

billyvg commented on src/sentry/static/sentry/app/components/events/contextSummary.jsx:
```diff
@@ -133,59 +159,32 @@ class EventContextSummary extends React.Component {
 
+    // Add contents in the declared order until we have at least MIN_CONTEXTS
+    // contexts in our list.
+    children = KNOWN_CONTEXTS.map(({key, Component, ...props}, index) => {
```
we could avoid this `map` if children already meets `MIN_CONTEXTS`

billyvg commented on src/sentry/static/sentry/app/components/events/contextSummary.jsx:
```diff
@@ -107,23 +107,49 @@ class DeviceSummary extends React.Component {
         <span className="context-item-icon" />
-        <h3>{deviceNameMapper(data.model)}</h3>
-        <p>{data.arch || data.model_id || ''}</p>
+        <h3>{data.model ? deviceNameMapper(data.model) : 'Unknown Device'}</h3>
```
`t()`

billyvg commented on src/sentry/static/sentry/app/components/events/contextSummary.jsx:
```diff
@@ -107,23 +107,49 @@ class DeviceSummary extends React.Component {
+const MAX_CONTEXTS = 4;
+const KNOWN_CONTEXTS = [
+  {key: 'user', Component: UserSummary},
+  {key: 'browser', Component: GenericSummary, unknownTitle: 'Unknown Browser'},
```
`t()`

jan-auer commented:
Updated based on review notes. Will add a test tomorrow. 

@billyvg the changed order in Percy is a side effect of unifying all contexts, but it is certainly not undesirable.

billyvg commented:
Shouldn't this [diff](https://percy.io/getsentry/sentry/builds/658136/view/34309994/1280?mode=head) have a browser + platform?

jan-auer commented:
That diff should not be there alltogether... See [here](https://github.com/getsentry/sentry/pull/8051/files#diff-af6a04783ec9ae8b4cff8f0437a50e46R174)

billyvg commented:
Ah is that just incomplete mock data? Or am I misunderstanding the relation between contexts and tags?

jan-auer commented:
@billyvg I think you might have been looking at an outdated percy diff. Updated tests and added more snapshots for different kinds of issues that might be affected by this. To me, they look good now, but I'll let you approve them on percy.
